source/custom/nidcpower_service.custom.cpp

```diff
@@ -1,5 +1,45 @@
 #include <nidcpower/nidcpower_service.h>
 
 namespace nidcpower_grpc {
+  //---------------------------------------------------------------------
+  //---------------------------------------------------------------------
+  ::grpc::Status NiDCPowerService::FetchMultiple(::grpc::ServerContext* context, const FetchMultipleRequest* request, FetchMultipleResponse* response)
+  {
+    if (context->IsCancelled()) {
+      return ::grpc::Status::CANCELLED;
+    }
+    try {
+      auto vi_grpc_session = request->vi();
+      ViSession vi = session_repository_->access_session(vi_grpc_session.id(), vi_grpc_session.name());
+      ViConstString channel_name = request->channel_name().c_str();
+      ViReal64 timeout = request->timeout();
+      ViInt32 count = request->count();
+      response->mutable_voltage_measurements()->Resize(count, 0);
+      ViReal64* voltage_measurements = response->mutable_voltage_measurements()->mutable_data();
+      response->mutable_current_measurements()->Resize(count, 0);
+      ViReal64* current_measurements = response->mutable_current_measurements()->mutable_data();
+      response->mutable_in_compliance()->Resize(count, 0);
```

Comment:
The changes for supporting ViBoolean array are from line 22 to 28 here, and the Copy function below.

As per IVisaType.h, ViBoolean is ViUInt16, which is `unsigned short`. Boolean equivalent in grpc is `bool`. We can't reinterpret_cast bool* as ViBoolean* since bool is 1 byte and ViBoolean (unsigned short) is 2 bytes. Solution we have is, we will allocate array of ViBoolean here and pass that array to FetchMultiple API. After `library_.FetchMultiple` returns, the values are copied to bool[] in response object. Implicit typecast happening while adding individual items to bool array should behave properly.

Reply:
FYI @kruthi-NI @Akshata440 @ganeshnie

Reply:
We shouldn't custom implement `ViBoolean[]`, but we should add codegen support for it. We just recently added code to support the `ViAttr` type, which is also a 16-bit integer value. We should be able to just add the `ViBoolean` type to those conditions in the mako scripts.

Reply:
Apologies I was not clear on the question. The question was more of, is it ok to create that extra copy of data and the effect copying of that data might have on performance of the API. For small arrays the effect might not be visible, but if we are trying to send large chunk of data across - this could have significant hit on performance of the API.

I will try out modifying mako script for auto-generation today.

Reply:
We have to make a copy of the 16-bit-integer arrays to 32-bit-integer. At this point we are not worrying about performance until we get feedback from customers that the performance is hindering them.